File: ETC/1717/a.cpp
#include <bits/stdc++.h>
#define fastIO ios::sync_with_stdio(0), cin.tie(0), cout.tie(0);

using namespace std;

int n,m;
int c,a,b;
vector<int> parent;

void init() {
    cin>>n>>m;
    for(int i=0;i<=n;i++){
        parent.push_back(i);
    }
}

int findParent(int x){
    if(parent[x]!=x){
        parent[x]=findParent(parent[x]);
    }
    return parent[x];
}

void go() {
    for(int i=0;i<m;i++){
        cin>>c>>a>>b;

        int parent_a=findParent(a);
        int parent_b=findParent(b);

        if(c==0){
            if(parent_a>parent_b){
                parent[parent_a]=parent_b;
            }else{
                parent[parent_b]=parent_a;
            }
        }else{
            cout<<(parent_a==parent_b ? "YES\n" : "NO\n");
        }
    }
}

int main() {
  fastIO;
  init();
  go();
  return 0;
}

File: ETC/1717/a.py
import sys

input=sys.stdin.readline

n,m=list(map(int,input().split()))

sys.setrecursionlimit(1000000)

parent=[0]*(n+1)
for i in range(n+1):
    parent[i]=i

def findParent(x):
    if parent[x]==x:
        parent[x]=findParent(parent[x])
    return parent[x]

for _ in range(m):
    cmd,a,b=list(map(int,input().split()))

    parent_a=findParent(a)
    parent_b=findParent(b)

    if cmd==0:
        if parent_a>parent_b:
            parent[parent_a]=parent_b
        else:
            parent[parent_b]=parent_a
    else:
        if parent_a==parent_b:
            print("YES")
        else:
            print("NO")